nevergrad/optimization/base.py

```diff
@@ -391,33 +391,36 @@ def ask(self, cheap_constraints_checker: Any = None) -> Candidate:
             The candidate to try on the objective function. Candidates have field `args` and `kwargs` which can be directly used
             on the function (`objective_function(*candidate.args, **candidate.kwargs)`).
         """
-        num_tries = 0
-        while True:  # Until we satisfy cheap constraints.
+        for callback in self._callbacks.get("ask", []):
+            callback(self)
+        MAX_TENTATIVES = 1000
+        current_num_ask = self.num_ask
+        for k in range(MAX_TENTATIVES):
             # call callbacks for logging etc...
-            for callback in self._callbacks.get("ask", []):
-                callback(self)
+            is_suggestion = False
             if self._suggestions:
+                is_suggestion = True
                 candidate = self._suggestions.pop()
             else:
                 candidate = self._internal_ask_candidate()
                 # only register actual asked points
-                if candidate.uid in self._asked:
-                    raise RuntimeError(
-                        "Cannot submit the same candidate twice: please recreate a new candidate from data.\n"
-                        "This is to make sure that stochastic instrumentations are resampled."
-                    )
-                self._asked.add(candidate.uid)
-            assert candidate is not None, f"{self.__class__.__name__}._internal_ask method returned None instead of a point."
-            if not cheap_constraints_checker or cheap_constraints_checker(candidate.data):
```

Comment:
better not check the constraint multiple times even if cheap. Also, it must use args and kwargs, not data which is internal